core/src/data/rasterSource.cpp

```diff
@@ -77,100 +91,133 @@ RasterSource::RasterSource(const std::string& _name, std::unique_ptr<DataSource>
+        m_tileData->layers.emplace_back("");
+        m_tileData->layers.back().features.push_back(rasterFeature);
     }
+}
+
+std::unique_ptr<Texture> RasterSource::createTexture(TileID _tile, const std::vector<char>& _rawTileData) {
+    if (_rawTileData.empty()) { return nullptr; }
 
     auto data = reinterpret_cast<const uint8_t*>(_rawTileData.data());
     auto length = _rawTileData.size();
 
-    std::shared_ptr<Texture> texture(new Texture(data, length, m_texOptions),
-                                     [c = std::weak_ptr<Cache>(m_textures), _tile](auto t) {
-                                         if (auto cache = c.lock()) { cache->erase(_tile); }
-                                         delete t;
-                                     });
-
-    return texture;
+    return std::make_unique<Texture>(data, length, m_texOptions);
 }
 
 void RasterSource::loadTileData(std::shared_ptr<TileTask> _task, TileTaskCb _cb) {
     // TODO, remove this
     // Overwrite cb to set empty texture on failure
     TileTaskCb cb{[this, _cb](std::shared_ptr<TileTask> _task) {
-
-            if (!_task->hasData()) {
-                auto& task = static_cast<RasterTileTask&>(*_task);
-                task.m_texture = m_emptyTexture;
-            }
-            _cb.func(_task);
-        }};
-
+        if (!_task->hasData()) {
+            auto& task = static_cast<RasterTileTask&>(*_task);
+            task.raster = std::make_unique<Raster>(task.tileId(), m_emptyTexture);
+        }
+        _cb.func(_task);
+    }};
     TileSource::loadTileData(_task, cb);
 }
 
 std::shared_ptr<TileData> RasterSource::parse(const TileTask& _task) const {
+    assert(false);
+    return nullptr;
+}
 
-    std::shared_ptr<TileData> tileData = std::make_shared<TileData>();
+void RasterSource::addRasterTask(TileTask& _task) {
 
-    Feature rasterFeature;
-    rasterFeature.geometryType = GeometryType::polygons;
-    rasterFeature.polygons = { { {
-                                         {0.0f, 0.0f},
-                                         {1.0f, 0.0f},
-                                         {1.0f, 1.0f},
-                                         {0.0f, 1.0f},
-                                         {0.0f, 0.0f}
-                                 } } };
-    rasterFeature.props = Properties();
+    TileID subTileID = _task.tileId();
 
-    tileData->layers.emplace_back("");
-    tileData->layers.back().features.push_back(rasterFeature);
-    return tileData;
+    // get tile for lower zoom if we are past max zoom
+    if (subTileID.z > maxZoom()) {
+        subTileID = subTileID.withMaxSourceZoom(maxZoom());
+    }
 
-}
+    auto rasterTask = createRasterTask(subTileID, true);
+
+    _task.subTasks().push_back(rasterTask);
 
-std::shared_ptr<TileTask> RasterSource::createTask(TileID _tileId, int _subTask) {
-    auto task = std::make_shared<RasterTileTask>(_tileId, shared_from_this(), _subTask);
+    // Needed? Do we support recursive raster-source inclusion?
```

Comment:
We could have a normal raster source on a raw raster/image tile source. So possible to have at least 1 more level.
Can you remove this comment, and maybe add above reasoning on when can we have raster source dependent on another raster source?

Reply:
Ahh... I see in this case `TileTask` is our 1st raster source. That said, we can remove this, do not know of any use case for the same.

Reply:
cc. @bcamper